File: apple/SlideKit/Sources/SlideKit/Fusion.swift
import CoreGraphics
import Foundation
import Vision

/// Align and exposure-fuse several scans of one slide — Mertens et al., as OpenCV's MergeMertens
/// does it with its defaults (contrast and saturation weights 1, well-exposedness 0).
///
/// Memory: works one scan at a time against the result pyramid, so the peak is roughly the result
/// pyramid plus one scan's pyramid and weights (~1 GB for a 22 MP bracket in Float). The roadmap's
/// half-precision, tiled GPU version is the next step for older iPads.
public enum Fusion {
    public static func fuse(_ images: [RGBImage], align shouldAlign: Bool = true) -> RGBImage {
        fuse(sources: images, align: shouldAlign)
    }

    /// Mertens over any row sources, one colour channel at a time: the peak is one channel's
    /// pyramids (result, scan, weights) plus the output, never a whole RGB pyramid per scan. For a
    /// 20 MP bracket read from 8-bit scans that is well under 1 GB.
    public static func fuse(sources input: [any RowSource], align shouldAlign: Bool = true) -> RGBImage {
        guard input.count > 1 else { return input.first?.rgbImage() ?? RGBImage(width: 1, height: 1) }
        let w = input.map(\.width).min()!, h = input.map(\.height).min()!
        let cropped: [any RowSource] = input.map { $0.width == w && $0.height == h ? $0 : CroppedSource(base: $0, width: w, height: h) }
        let sources = shouldAlign ? align(cropped) : cropped

        var weightSum = Plane(width: w, height: h)
        for src in sources {
            let wt = weights(src)
            weightSum.data.withUnsafeMutableBufferPointer { s in
                wt.data.withUnsafeBufferPointer { v in let s = s; parallelFor(w * h, minChunk: 4096) { r in for i in r { s[i] += v[i] } } }
            }
        }
        let levels = Int(log(Double(min(w, h))) / log(2.0))
        var out = RGBImage(width: w, height: h)
        for c in 0..<3 {
            var result: [Plane] = []
            for src in sources {
                // the weights are cheap to recompute; keeping three weight pyramids is not
                var wt = weights(src)
                wt.data.withUnsafeMutableBufferPointer { v in
                    weightSum.data.withUnsafeBufferPointer { s in let v = v; parallelFor(w * h, minChunk: 4096) { r in for i in r { v[i] /= s[i] } } }
                }
                let wPyr = Pyramid.gaussian(wt, levels: levels)
                wt = Plane(width: 0, height: 0)
                var lap = Pyramid.gaussian(channel(src, c), levels: levels)
                for l in 0..<levels { subtract(&lap[l], Pyramid.up(lap[l + 1], width: lap[l].width, height: lap[l].height)) }
                for l in 0...levels {
                    multiply(&lap[l], wPyr[l])
                    if result.count <= l { result.append(lap[l]) } else { add(&result[l], lap[l]) }
                    lap[l] = Plane(width: 0, height: 0)
                }
            }
            for l in stride(from: levels, to: 0, by: -1) {
                add(&result[l - 1], Pyramid.up(result[l], width: result[l - 1].width, height: result[l - 1].height))
                result.removeLast()
            }
            let plane = result[0]
            out.data.withUnsafeMutableBufferPointer { o in
                plane.data.withUnsafeBufferPointer { p in
                    let o = o
                    parallelFor(w * h, minChunk: 4096) { r in for i in r { o[i * 3 + c] = min(1, max(0, p[i])) } }
                }
            }
        }
        return out
    }

    /// contrast × saturation + 1e-12, per pixel.
    static func weights(_ src: any RowSource) -> Plane {
        let w = src.width, h = src.height
        var gray = Plane(width: w, height: h), sat = Plane(width: w, height: h)
        gray.data.withUnsafeMutableBufferPointer { g in
            sat.data.withUnsafeMutableBufferPointer { s in
                let g = g, s = s
                src.forEachRow { y, row in
                    for x in 0..<w {
                        let r = row[x * 3], gg = row[x * 3 + 1], b = row[x * 3 + 2]
                        g[y * w + x] = 0.299 * r + 0.587 * gg + 0.114 * b
                        let m = (r + gg + b) / 3
                        s[y * w + x] = ((r - m) * (r - m) + (gg - m) * (gg - m) + (b - m) * (b - m)).squareRoot()
                    }
                }
            }
        }
        let contrast = Filters.laplacian(gray)
        gray = Plane(width: 0, height: 0)
        sat.data.withUnsafeMutableBufferPointer { s in
            contrast.data.withUnsafeBufferPointer { c in let s = s; parallelFor(w * h, minChunk: 4096) { r in for i in r { s[i] = abs(c[i]) * s[i] + 1e-12 } } }
        }
        return sat
    }

    static func channel(_ src: any RowSource, _ c: Int) -> Plane {
        let w = src.width
        var out = Plane(width: w, height: src.height)
        out.data.withUnsafeMutableBufferPointer { o in
            let o = o
            src.forEachRow { y, row in for x in 0..<w { o[y * w + x] = row[x * 3 + c] } }
        }
        return out
    }

    static func subtract(_ a: inout Plane, _ b: Plane) { combine(&a, b) { $0 - $1 } }
    static func add(_ a: inout Plane, _ b: Plane) { combine(&a, b) { $0 + $1 } }
    static func multiply(_ a: inout Plane, _ b: Plane) { combine(&a, b) { $0 * $1 } }
    @inline(__always) static func combine(_ a: inout Plane, _ b: Plane, _ f: (Float, Float) -> Float) {
        let n = a.data.count
        a.data.withUnsafeMutableBufferPointer { x in
            b.data.withUnsafeBufferPointer { y in let x = x; parallelFor(n, minChunk: 8192) { r in for i in r { x[i] = f(x[i], y[i]) } } }
        }
    }

    // MARK: alignment

    /// Shift every scan onto the first. The scanner holds the slide still, but a hand-bracketed stack
    /// can move by a few pixels between presses. Registration runs on small copies and is scaled up;
    /// the shift is applied while reading, not by copying the scan.
    static func align(_ sources: [any RowSource]) -> [any RowSource] {
        guard sources.count > 1 else { return sources }
        let ref = sources[0]
        let small = ref.thumbnail(maxEdge: 1200)
        let scale = Double(small.width) / Double(ref.width)
        let refCG = ImageFile.cgImage(small)
        return [ref] + sources.dropFirst().map { src -> any RowSource in
            let request = VNTranslationalImageRegistrationRequest(targetedCGImage: ImageFile.cgImage(src.thumbnail(maxEdge: 1200)))
            try? VNImageRequestHandler(cgImage: refCG, options: [:]).perform([request])
            guard let t = request.results?.first?.alignmentTransform else { return src }
            let dx = Int((t.tx / scale).rounded()), dy = Int((-t.ty / scale).rounded())  // Vision's y points up
            guard dx != 0 || dy != 0, abs(dx) < src.width / 10, abs(dy) < src.height / 10 else { return src }
            return ShiftedSource(base: src, dx: dx, dy: dy)
        }
    }

    static func shifted(_ im: RGBImage, dx: Int, dy: Int) -> RGBImage { ShiftedSource(base: im, dx: dx, dy: dy).rgbImage() }
}

/// OpenCV's pyrDown / pyrUp: the 5-tap [1 4 6 4 1] / 16 kernel, BORDER_REFLECT_101.
enum Pyramid {
    static let k: [Float] = [1, 4, 6, 4, 1].map { $0 / 16 }

    static func gaussian(_ p: Plane, levels: Int) -> [Plane] {
        var out = [p]
        for _ in 0..<levels { out.append(down(out.last!)) }
        return out
    }
    /// Blur with [1 4 6 4 1]/16 and keep every other pixel — computing only the kept samples.
    static func down(_ p: Plane) -> Plane {
        let w = p.width, h = p.height, ow = (w + 1) / 2, oh = (h + 1) / 2
        var tmp = [Float](repeating: 0, count: w * oh)   // vertical pass on the even rows
        p.data.withUnsafeBufferPointer { s in
            tmp.withUnsafeMutableBufferPointer { t in
                let t = t
                parallelFor(oh) { rows in
                    for oy in rows {
                        let y = oy * 2
                        let r = (-2...2).map { reflect101(y + $0, h) * w }
                        for x in 0..<w {
                            t[oy * w + x] = (s[r[0] + x] + s[r[4] + x]) * k[0] + (s[r[1] + x] + s[r[3] + x]) * k[1] + s[r[2] + x] * k[2]
                        }
                    }
                }
            }
        }
        var out = Plane(width: ow, height: oh)
        tmp.withUnsafeBufferPointer { t in
            out.data.withUnsafeMutableBufferPointer { o in
                let o = o
                parallelFor(oh) { rows in
                    for oy in rows {
                        let row = oy * w
                        for ox in 0..<ow {
                            let x = ox * 2
                            o[oy * ow + ox] = (t[row + reflect101(x - 2, w)] + t[row + reflect101(x + 2, w)]) * k[0]
                                + (t[row + reflect101(x - 1, w)] + t[row + reflect101(x + 1, w)]) * k[1] + t[row + x] * k[2]
                        }
                    }
                }
            }
        }
        return out
    }

    /// Upsample to (width, height) exactly as cv::pyrUp does: per axis, even outputs are
    /// (s[i-1] + 6 s[i] + s[i+1]) / 8 and odd ones (s[i] + s[i+1]) / 2, reflecting at the start
    /// (s[-1] = s[1]) but repeating at the end (s[n] = s[n-1]). The borders matter: coarse levels
    /// are only a few pixels wide and carry the picture's overall colour.
    static func up(_ p: Plane, width: Int, height: Int) -> Plane {
        func axis(_ n: Int, _ outN: Int) -> [[(Int, Float)]] {
            (0..<outN).map { d in
                let i = min(n - 1, d / 2)
                let prev = i == 0 ? min(1, n - 1) : i - 1, next = min(n - 1, i + 1)
                let even: [(Int, Float)] = [(prev, 0.125), (i, 0.75), (next, 0.125)], odd: [(Int, Float)] = [(i, 0.5), (next, 0.5)]
                return d % 2 == 0 ? even : odd
            }
        }
        let xs = axis(p.width, width), ys = axis(p.height, height)
        var tmp = [Float](repeating: 0, count: width * p.height)
        p.data.withUnsafeBufferPointer { s in
            tmp.withUnsafeMutableBufferPointer { t in
                let t = t
                parallelFor(p.height) { rows in
                    for y in rows {
                        for (x, taps) in xs.enumerated() {
                            var acc: Float = 0
                            for (i, w) in taps { acc += s[y * p.width + i] * w }
                            t[y * width + x] = acc
                        }
                    }
                }
            }
        }
        var out = Plane(width: width, height: height)
        tmp.withUnsafeBufferPointer { t in
            out.data.withUnsafeMutableBufferPointer { o in
                let o = o
                parallelFor(height) { rows in
                    for y in rows {
                        let taps = ys[y]
                        for x in 0..<width {
                            var acc: Float = 0
                            for (i, w) in taps { acc += t[i * width + x] * w }
                            o[y * width + x] = acc
                        }
                    }
                }
            }
        }
        return out
    }
}

File: frontend/src/standalone/fusion.ts
// Align and exposure-fuse several scans of one slide (imaging.fuse): Mertens as OpenCV's
// MergeMertens does it with its defaults, ported from apple/SlideKit's Fusion.swift, and a
// median-threshold-bitmap alignment in the spirit of OpenCV's AlignMTB.
import { plane, reflect101, type Plane, type RGB } from "./pixels";

/** Hands out one row of float RGB at a time: a decoded 8-bit scan, a shifted view, or an RGB. */
export type RowSource = { width: number; height: number; row(y: number, out: Float32Array): void };

export const rgbSource = (a: RGB): RowSource => ({
  width: a.width,
  height: a.height,
  row: (y, out) => out.set(a.data.subarray(y * a.width * 3, (y + 1) * a.width * 3)),
});

/** 8-bit RGBA (ImageData) — a quarter of the memory of float RGB, for full-resolution work. */
export const rgba8Source = (width: number, height: number, bytes: Uint8ClampedArray): RowSource => ({
  width,
  height,
  row: (y, out) => {
    const base = y * width * 4;
    for (let x = 0; x < width; x++) {
      out[x * 3] = bytes[base + x * 4] / 255;
      out[x * 3 + 1] = bytes[base + x * 4 + 1] / 255;
      out[x * 3 + 2] = bytes[base + x * 4 + 2] / 255;
    }
  },
});

/** The top-left width × height of a larger source (brackets whose scans differ by a pixel). */
const croppedSource = (base: RowSource, width: number, height: number): RowSource => {
  const tmp = new Float32Array(base.width * 3);
  return {
    width,
    height,
    row: (y, out) => {
      base.row(y, tmp);
      out.set(tmp.subarray(0, width * 3));
    },
  };
};

/** A source moved by (dx, dy) pixels, edges repeated — alignment without a copy. */
export const shiftedSource = (base: RowSource, dx: number, dy: number): RowSource => {
  const tmp = new Float32Array(base.width * 3);
  const w = base.width;
  return {
    width: w,
    height: base.height,
    row: (y, out) => {
      base.row(Math.min(base.height - 1, Math.max(0, y - dy)), tmp);
      for (let x = 0; x < w; x++) {
        const sx = Math.min(w - 1, Math.max(0, x - dx)) * 3;
        out[x * 3] = tmp[sx];
        out[x * 3 + 1] = tmp[sx + 1];
        out[x * 3 + 2] = tmp[sx + 2];
      }
    },
  };
};

export function materialise(src: RowSource): RGB {
  const out = new Float32Array(src.width * src.height * 3);
  const buf = new Float32Array(src.width * 3);
  for (let y = 0; y < src.height; y++) {
    src.row(y, buf);
    out.set(buf, y * src.width * 3);
  }
  return { width: src.width, height: src.height, data: out };
}

// ------------------------------------------------------------------ Mertens

export function fuse(images: RGB[], align = true): RGB {
  return fuseSources(images.map(rgbSource), align);
}

/**
 * Mertens over row sources, one colour channel at a time: the peak is one channel's pyramids
 * (result, scan, weights) plus the output, never a whole RGB pyramid per scan.
 */
export function fuseSources(input: RowSource[], align = true): RGB {
  if (input.length === 1) return materialise(input[0]);
  const w = Math.min(...input.map((s) => s.width));
  const h = Math.min(...input.map((s) => s.height));
  const same = input.map((s) => (s.width === w && s.height === h ? s : croppedSource(s, w, h)));
  const sources = align ? alignSources(same) : same;

  const weightSum = plane(w, h);
  for (const src of sources) {
    const wt = weights(src);
    for (let i = 0; i < w * h; i++) weightSum.data[i] += wt.data[i];
  }
  const levels = Math.trunc(Math.log(Math.min(w, h)) / Math.log(2));
  const out = new Float32Array(w * h * 3);
  for (let c = 0; c < 3; c++) {
    const result: Plane[] = [];
    for (const src of sources) {
      // the weights are cheap to recompute; keeping three weight pyramids is not
      const wt = weights(src);
      for (let i = 0; i < w * h; i++) wt.data[i] /= weightSum.data[i];
      const wPyr = gaussianPyramid(wt, levels);
      const lap = gaussianPyramid(channelOf(src, c), levels);
      for (let l = 0; l < levels; l++) combine(lap[l], pyrUp(lap[l + 1], lap[l].width, lap[l].height), (a, b) => a - b);
      for (let l = 0; l <= levels; l++) {
        combine(lap[l], wPyr[l], (a, b) => a * b);
        if (result.length <= l) result.push(lap[l]);
        else combine(result[l], lap[l], (a, b) => a + b);
      }
    }
    for (let l = levels; l > 0; l--) {
      combine(result[l - 1], pyrUp(result[l], result[l - 1].width, result[l - 1].height), (a, b) => a + b);
      result.pop();
    }
    const p = result[0].data;
    for (let i = 0; i < w * h; i++) out[i * 3 + c] = Math.min(1, Math.max(0, p[i]));
  }
  return { width: w, height: h, data: out };
}

function combine(a: Plane, b: Plane, f: (x: number, y: number) => number) {
  const x = a.data;
  const y = b.data;
  for (let i = 0; i < x.length; i++) x[i] = f(x[i], y[i]);
}

/** contrast × saturation + 1e-12, per pixel. */
function weights(src: RowSource): Plane {
  const { width: w, height: h } = src;
  const g = plane(w, h);
  const sat = plane(w, h);
  const row = new Float32Array(w * 3);
  for (let y = 0; y < h; y++) {
    src.row(y, row);
    for (let x = 0; x < w; x++) {
      const r = row[x * 3];
      const gg = row[x * 3 + 1];
      const b = row[x * 3 + 2];
      g.data[y * w + x] = 0.299 * r + 0.587 * gg + 0.114 * b;
      const m = (r + gg + b) / 3;
      sat.data[y * w + x] = Math.sqrt((r - m) * (r - m) + (gg - m) * (gg - m) + (b - m) * (b - m));
    }
  }
  const lapl = laplacianOf(g);
  for (let i = 0; i < w * h; i++) sat.data[i] = Math.abs(lapl.data[i]) * sat.data[i] + 1e-12;
  return sat;
}

function laplacianOf(p: Plane): Plane {
  const { width: w, height: h } = p;
  const out = plane(w, h);
  for (let y = 0; y < h; y++) {
    const yu = reflect101(y - 1, h) * w;
    const yd = reflect101(y + 1, h) * w;
    const r = y * w;
    for (let x = 0; x < w; x++)
      out.data[r + x] =
        p.data[yu + x] +
        p.data[yd + x] +
        p.data[r + reflect101(x - 1, w)] +
        p.data[r + reflect101(x + 1, w)] -
        4 * p.data[r + x];
  }
  return out;
}

function channelOf(src: RowSource, c: number): Plane {
  const out = plane(src.width, src.height);
  const row = new Float32Array(src.width * 3);
  for (let y = 0; y < src.height; y++) {
    src.row(y, row);
    for (let x = 0; x < src.width; x++) out.data[y * src.width + x] = row[x * 3 + c];
  }
  return out;
}

const K = [1 / 16, 4 / 16, 6 / 16];

function gaussianPyramid(p: Plane, levels: number): Plane[] {
  const out = [p];
  for (let l = 0; l < levels; l++) out.push(pyrDown(out[l]));
  return out;
}

/** cv::pyrDown: blur with [1 4 6 4 1]/16 and keep every other pixel (BORDER_REFLECT_101). */
function pyrDown(p: Plane): Plane {
  const { width: w, height: h } = p;
  const ow = (w + 1) >> 1;
  const oh = (h + 1) >> 1;
  const s = p.data;
  const tmp = new Float32Array(w * oh);
  for (let oy = 0; oy < oh; oy++) {
    const y = oy * 2;
    const r0 = reflect101(y - 2, h) * w;
    const r1 = reflect101(y - 1, h) * w;
    const r2 = y * w;
    const r3 = reflect101(y + 1, h) * w;
    const r4 = reflect101(y + 2, h) * w;
    for (let x = 0; x < w; x++)
      tmp[oy * w + x] = (s[r0 + x] + s[r4 + x]) * K[0] + (s[r1 + x] + s[r3 + x]) * K[1] + s[r2 + x] * K[2];
  }
  const out = plane(ow, oh);
  for (let oy = 0; oy < oh; oy++) {
    const row = oy * w;
    for (let ox = 0; ox < ow; ox++) {
      const x = ox * 2;
      out.data[oy * ow + ox] =
        (tmp[row + reflect101(x - 2, w)] + tmp[row + reflect101(x + 2, w)]) * K[0] +
        (tmp[row + reflect101(x - 1, w)] + tmp[row + reflect101(x + 1, w)]) * K[1] +
        tmp[row + x] * K[2];
    }
  }
  return out;
}

/**
 * cv::pyrUp to (width, height): per axis, even outputs are (s[i-1] + 6 s[i] + s[i+1]) / 8 and odd
 * ones (s[i] + s[i+1]) / 2, reflecting at the start but repeating at the end.
 */
function pyrUp(p: Plane, width: number, height: number): Plane {
  const axis = (n: number, outN: number) => {
    const a = new Int32Array(outN * 3);
    const wt = new Float32Array(outN * 3);
    for (let d = 0; d < outN; d++) {
      const i = Math.min(n - 1, d >> 1);
      const prev = i === 0 ? Math.min(1, n - 1) : i - 1;
      const next = Math.min(n - 1, i + 1);
      if (d % 2 === 0) (a.set([prev, i, next], d * 3), wt.set([0.125, 0.75, 0.125], d * 3));
      else (a.set([i, next, next], d * 3), wt.set([0.5, 0.5, 0], d * 3));
    }
    return { a, wt };
  };
  const xs = axis(p.width, width);
  const ys = axis(p.height, height);
  const tmp = new Float32Array(width * p.height);
  for (let y = 0; y < p.height; y++) {
    const r = y * p.width;
    for (let x = 0; x < width; x++) {
      const k = x * 3;
      tmp[y * width + x] =
        p.data[r + xs.a[k]] * xs.wt[k] +
        p.data[r + xs.a[k + 1]] * xs.wt[k + 1] +
        p.data[r + xs.a[k + 2]] * xs.wt[k + 2];
    }
  }
  const out = plane(width, height);
  for (let y = 0; y < height; y++) {
    const k = y * 3;
    const r0 = ys.a[k] * width;
    const r1 = ys.a[k + 1] * width;
    const r2 = ys.a[k + 2] * width;
    const w0 = ys.wt[k];
    const w1 = ys.wt[k + 1];
    const w2 = ys.wt[k + 2];
    for (let x = 0; x < width; x++) out.data[y * width + x] = tmp[r0 + x] * w0 + tmp[r1 + x] * w1 + tmp[r2 + x] * w2;
  }
  return out;
}

// ------------------------------------------------------------------ alignment

/** Box-averaged grey copy with the longer edge at most maxEdge, and the factor it shrank by. */
function smallGray(src: RowSource, maxEdge: number): { p: Plane; f: number } {
  const f = Math.max(1, Math.ceil(Math.max(src.width, src.height) / maxEdge));
  const ow = Math.max(1, Math.trunc(src.width / f));
  const oh = Math.max(1, Math.trunc(src.height / f));
  const out = plane(ow, oh);
  const row = new Float32Array(src.width * 3);
  for (let oy = 0; oy < oh; oy++) {
    for (let sy = oy * f; sy < oy * f + f; sy++) {
      src.row(sy, row);
      for (let ox = 0; ox < ow; ox++)
        for (let sx = ox * f; sx < ox * f + f; sx++)
          out.data[oy * ow + ox] += 0.299 * row[sx * 3] + 0.587 * row[sx * 3 + 1] + 0.114 * row[sx * 3 + 2];
    }
    for (let ox = 0; ox < ow; ox++) out.data[oy * ow + ox] /= f * f;
  }
  return { p: out, f };
}

const half = (p: Plane): Plane => {
  const ow = p.width >> 1;
  const oh = p.height >> 1;
  const out = plane(ow, oh);
  for (let y = 0; y < oh; y++)
    for (let x = 0; x < ow; x++) {
      const i = 2 * y * p.width + 2 * x;
      out.data[y * ow + x] = (p.data[i] + p.data[i + 1] + p.data[i + p.width] + p.data[i + p.width + 1]) / 4;
    }
  return out;
};

/** Median threshold bitmap and its exclusion mask (pixels too close to the median to trust). */
function mtb(p: Plane, exclude = 4 / 255) {
  const sorted = Float32Array.from(p.data).sort();
  const med = sorted[sorted.length >> 1];
  const tb = new Uint8Array(p.data.length);
  const eb = new Uint8Array(p.data.length);
  for (let i = 0; i < p.data.length; i++) {
    tb[i] = p.data[i] > med ? 1 : 0;
    eb[i] = Math.abs(p.data[i] - med) > exclude ? 1 : 0;
  }
  return { tb, eb, w: p.width, h: p.height };
}

type Bitmap = ReturnType<typeof mtb>;

function mtbError(a: Bitmap, b: Bitmap, dx: number, dy: number): number {
  let err = 0;
  for (let y = Math.max(0, dy); y < Math.min(a.h, b.h + dy); y++) {
    const ra = y * a.w;
    const rb = (y - dy) * b.w;
    for (let x = Math.max(0, dx); x < Math.min(a.w, b.w + dx); x++) {
      const i = ra + x;
      const j = rb + x - dx;
      err += (a.tb[i] ^ b.tb[j]) & a.eb[i] & b.eb[j];
    }
  }
  return err;
}

/** The (dx, dy) that moves `b` onto `a`, coarse to fine over six levels (Ward's MTB). */
export function mtbShift(a: Plane, b: Plane, maxBits = 6): [number, number] {
  const pa: Plane[] = [a];
  const pb: Plane[] = [b];
  for (let l = 1; l < maxBits && Math.min(pa[l - 1].width, pa[l - 1].height) >= 16; l++) {
    pa.push(half(pa[l - 1]));
    pb.push(half(pb[l - 1]));
  }
  let sx = 0;
  let sy = 0;
  for (let l = pa.length - 1; l >= 0; l--) {
    sx *= 2;
    sy *= 2;
    const ba = mtb(pa[l]);
    const bb = mtb(pb[l]);
    let best = [sx, sy];
    let bestErr = mtbError(ba, bb, sx, sy);
    for (let j = -1; j <= 1; j++)
      for (let i = -1; i <= 1; i++) {
        if (!i && !j) continue;
        const e = mtbError(ba, bb, sx + i, sy + j);
        if (e < bestErr) ((bestErr = e), (best = [sx + i, sy + j]));
      }
    [sx, sy] = best;
  }
  return [sx, sy];
}

/**
 * Shift every scan onto the first. The scanner holds the slide still, but a hand-bracketed stack
 * can move by a few pixels between presses. Registration runs on small copies and is scaled up.
 */
export function alignSources(sources: RowSource[]): RowSource[] {
  if (sources.length < 2) return sources;
  const ref = smallGray(sources[0], 1200);
  return [
    sources[0],
    ...sources.slice(1).map((src) => {
      const s = smallGray(src, 1200);
      const [dx0, dy0] = mtbShift(ref.p, s.p);
      const dx = dx0 * ref.f;
      const dy = dy0 * ref.f;
      if ((!dx && !dy) || Math.abs(dx) >= src.width / 10 || Math.abs(dy) >= src.height / 10) return src;
      return shiftedSource(src, dx, dy);
    }),
  ];
}
